### PR TITLE
Re-implement info icon/help text popup

### LINKED ISSUE #79: Re-implement info icon/help text popup
We should implement a help text system similar to the system we had in the json-editor.

Supersedes #6

### PR DESCRIPTION
**Pull request by tulir**
Closes #79 

## Changes
* Create a tooltip Aurelia element
* Add support for a info button with a tooltip to all fields
  * Uses the schema field `helpText`

### CODE REVIEW

File: res/info.svg [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
+<!-- Source: https://materialdesignicons.com/icon/information-outline -->
```
Comment (brylie) [inline res/info.svg]: Clean up commented code.

File: src/resources/elements/arrayfield.html [hunk 2]
```diff
@@ -1,6 +1,9 @@
 <template>
   <div class="array field c${columns} columns" if.bind="display">
-    <h3 click.delegate="toggleCollapse()" class="title" textcontent.bind="label"></h3>
+    <h3 click.delegate="toggleCollapse()" class="title">
+      ${label}
+      <tooltip if.bind="helpText" text.bind="helpText"><img src="res/info.svg" alt="Info"/></tooltip>
```
Comment (brylie) [inline src/resources/elements/arrayfield.html]: For readability, put nested child elements on own line.

File: src/resources/elements/objectfield.html [hunk 3]
```diff
@@ -5,7 +5,10 @@
         <compose containerless view-model.bind="childElem"></compose>
       </template>
     </legend>
-    <h3 click.delegate="toggleCollapse()" class="title" textcontent.bind="label"></h3>
+    <h3 click.delegate="toggleCollapse()" class="title">
+      ${label}
+      <tooltip if.bind="helpText" text.bind="helpText"><img src="res/info.svg" alt="Info"/></tooltip>
```
Comment (brylie) [inline src/resources/elements/objectfield.html]: For readability, put nested child elements on own line.

File: src/resources/elements/optionfield-checkbox.html [hunk 4]
```diff
@@ -1,6 +1,9 @@
 <template>
   <div class="checkbox option field c${columns} columns" if.bind="display">
-    <label textcontent.bind="label"></label>
+    <label for.bind="id">
+      ${label}
+      <tooltip if.bind="helpText" text.bind="helpText"><img src="res/info.svg" alt="Info"/></tooltip>
```
Comment (brylie) [inline src/resources/elements/optionfield-checkbox.html]: For readability, put nested child elements on own line.

File: src/resources/elements/optionfield-dropdown.html [hunk 5]
```diff
@@ -1,6 +1,9 @@
 <template>
   <div class="dropdown option field c${columns} columns" if.bind="display">
-    <label for.bind="id" textcontent.bind="label"></label>
+    <label for.bind="id">
+      ${label}
+      <tooltip if.bind="helpText" text.bind="helpText"><img src="res/info.svg" alt="Info"/></tooltip>
```
Comment (brylie) [inline src/resources/elements/optionfield-dropdown.html]: For readability, put nested child elements on own line.

File: src/resources/elements/textareafield.html [hunk 6]
```diff
@@ -1,6 +1,9 @@
 <template>
   <div class="textarea field c${columns} columns" if.bind="display">
-    <label for.bind="id" textcontent.bind="label"></label>
+    <label for.bind="id">
+      ${label}
+      <tooltip if.bind="helpText" text.bind="helpText"><img src="res/info.svg" alt="Info"/></tooltip>
```
Comment (brylie) [inline src/resources/elements/textareafield.html]: For readability, put nested child elements on own line.

File: src/resources/elements/textfield.html [hunk 7]
```diff
@@ -1,6 +1,9 @@
 <template>
   <div class="text field c${columns} columns" if.bind="display">
-    <label for.bind="id" textcontent.bind="label"></label>
+    <label for.bind="id">
+      ${label}
+      <tooltip if.bind="helpText" text.bind="helpText"><img src="res/info.svg" alt="Info"/></tooltip>
```
Comment (brylie) [inline src/resources/elements/textfield.html]: For readability, put nested child elements on own line.

File: src/style/index.scss [hunk 8]
```diff
@@ -76,6 +78,10 @@ h3 {
     &:empty {
       display: none;
     }
+
+    tooltip {
```
Comment (brylie) [inline src/style/index.scss]: Should this be `.tooltip`?

File: src/style/index.scss [hunk 9]
```diff
@@ -86,6 +92,13 @@ h3 {
     &:last-child {
       margin-bottom: 0;
     }
+
+
+
+    > tooltip {
```
Comment (brylie) [inline src/style/index.scss]: Should this be `.tooltip`?

**Comment by brylie** [comment]
On reading the code, it is well organized and easy to read. I have added a suggestion to several blocks of the HTML, regarding nesting child elements.

Also, please double check the scss file that the `tooltip` rules are correct.

File: src/resources/elements/arrayfield.html (hunk 2 above)
Comment (brylie) [inline src/resources/elements/arrayfield.html]: E.g.

```html
<tooltip if.bind="helpText" text.bind="helpText">
  <img src="res/info.svg" alt="Info"/>
</tooltip>
```

**Comment by brylie** [comment]
I am having difficulty testing the `helpText` attribute. I have placed a `helpText` attribute into several objects in a couple of schema files, but do not see any change in the user interface. How does the help text work?

File: src/style/index.scss (hunk 8 above)
Comment (tulir) [inline src/style/index.scss]: `tooltip` is a custom Aurelia element, so the current code should work.

File: res/info.svg (hunk 1 above)
Comment (tulir) [inline res/info.svg]: That's just the SVG file header and a comment telling where the image was downloaded from.

File: src/style/index.scss (hunk 8 above)
Comment (brylie) [inline src/style/index.scss]: Ah, I didn't catch that. Thanks for clarifying.

**Comment by brylie** [comment]
After starting `au run` on this branch, I get a whitescreen in the browser with the following console error:

```js
Unhandled rejection Error: Error invoking App. Check the inner error for details.
------------------------------------------------
Inner Error:
Message: this.legendChildren is undefined
```

**Comment by brylie** [comment]
Consider moving the help text icon closer to the field label. As it stands, right-alignment of the help text icon is a bit confusing:

screenshot_20170519_122159

Consider using [proximity](http://graphicdesign.spokanefalls.edu/tutorials/process/gestaltprinciples/gestaltprinc.htm#proximity) to show the relationship between the field (label) and it's help text.

**Comment by tulir** [comment]
The legendChildren error is probably the one fixed in #84. If it's not, I'll need the line numbers where it happens.

**Comment by brylie** [comment]
>  I'll need the line numbers where it happens.

[Here is the full stack trace](https://gist.github.com/brylie/a1cddc0bb4d1accbf465c9d358d840dc)

**Comment by tulir** [comment]
All the requested changes should be resolved now.

**Comment by brylie** [comment]
Help icons are still right aligned in their parent container, placed at a maximum distance away from the field label:

screenshot_20170523_100938

**Comment by tulir** [comment]
I positioned the info button to the right side of the labels and added better positioning for the popup.

**Comment by brylie** [comment]
@tulir consider adding the following style for the tooltip image:

```css
tooltip img {
    height: 0.8em;
    opacity: 0.5;
    vertical-align: super;
}
```

**Pull request merged**